File: CPP/155.cpp
class MinStack {
public:
    /** initialize your data structure here. */
    // Second will be used for tracking the min value
    stack<pair<int, int>> st;
    MinStack() {

    }

    void push(int x) {
        if (st.empty()) st.push({x,x});
        else {
            // Stack is not empty hence compare the value with current min
            if (st.top().second > x) st.push({x,x});
            else st.push({x, st.top().second});
        }
    }

    void pop() {
        st.pop();
    }

    int top() {
        return st.top().first;
    }

    int getMin() {
        return st.top().second;
    }
};

/**
 * Your MinStack object will be instantiated and called as such:
 * MinStack* obj = new MinStack();
 * obj->push(x);
 * obj->pop();
 * int param_3 = obj->top();
 * int param_4 = obj->getMin();
 */

File: Python/155.py
class MinStack:

    def __init__(self):
        """
        initialize your data structure here.
        """
        self.st = []


    def push(self, x: int) -> None:
        if len(self.st) == 0:
            self.st.append([x,x]);
        else:
            if self.st[-1][1] > x:
                self.st.append([x,x])
            else:
                self.st.append([x, self.st[-1][1]])

    def pop(self) -> None:
        self.st.pop()

    def top(self) -> int:
        return self.st[-1][0]

    def getMin(self) -> int:
        return self.st[-1][1]


# Your MinStack object will be instantiated and called as such:
# obj = MinStack()
# obj.push(x)
# obj.pop()
# param_3 = obj.top()
# param_4 = obj.getMin()
